syncer/streamer_controller.go

```diff
@@ -0,0 +1,297 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package syncer
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/pingcap/failpoint"
+	"github.com/siddontang/go-mysql/mysql"
+	"github.com/siddontang/go-mysql/replication"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/pkg/binlog"
+	tcontext "github.com/pingcap/dm/pkg/context"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/streamer"
+	"github.com/pingcap/dm/pkg/terror"
+	"github.com/pingcap/dm/pkg/utils"
+)
+
+var (
+	eventTimeout = 1 * time.Minute
+)
+
+// StreamerProducer provides the ability to generate binlog streamer by StartSync()
+// but go-mysql StartSync() returns (struct, err) rather than (interface, err)
+// And we can't simplely use StartSync() method in SteamerProducer
+// so use generateStreamer to wrap StartSync() method to make *BinlogSyncer and *BinlogReader in same interface
+// For other implementations who implement StreamerProducer and Streamer can easily take place of Syncer.streamProducer
+// For test is easy to mock
+type StreamerProducer interface {
+	generateStreamer(pos mysql.Position) (streamer.Streamer, error)
+}
+
+// Read local relay log
+type localBinlogReader struct {
+	reader *streamer.BinlogReader
+}
+
+func (l *localBinlogReader) generateStreamer(pos mysql.Position) (streamer.Streamer, error) {
+	return l.reader.StartSync(pos)
+}
+
+// Read remote binlog
+type remoteBinlogReader struct {
+	reader     *replication.BinlogSyncer
+	tctx       *tcontext.Context
+	EnableGTID bool
+}
+
+func (r *remoteBinlogReader) generateStreamer(pos mysql.Position) (streamer.Streamer, error) {
+	defer func() {
+		lastSlaveConnectionID := r.reader.LastConnectionID()
+		r.tctx.L().Info("last slave connection", zap.Uint32("connection ID", lastSlaveConnectionID))
+	}()
+
+	// FIXME: can enable GTID
+	if r.EnableGTID {
+		// NOTE: our (per-table based) checkpoint does not support GTID yet
+		return nil, terror.ErrSyncerUnitRemoteSteamerWithGTID.Generate()
+	}
+
+	// position's name may contain uuid, so need remove it
+	adjustedPos := binlog.AdjustPosition(pos)
```

Comment:
Do we need to check the possible existing UUID with the current upstream server's UUID?

Reply:
seems don't need to check uuid, but need to update the`syncCfg`, fix in 9ef6352f484fb9e7bb1a89a04be5b39573d5aa02

Reply:
check in 5513ad2